torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -192,6 +192,14 @@ void IndexLowering::handle(const TernaryOp* top) {
   GpuLower::current()->propagateExprInfo(top, back());
 }
 
+void IndexLowering::handle(const SelectOp* sop) {
+  const auto input = lowerSrcIndex(sop->input(0), sop->output(0));
+  const auto out = lowerDstIndex(sop->output(0));
+  pushBack(
+      IrBuilder::create<SelectOp>(out, input, sop->getDim(), sop->input(1)));
+  GpuLower::current()->propagateExprInfo(sop, back());
+}
```

Comment:
For `index_select`, I think we can do a `lowerSrcIndex` for the index tensor first, and then pass the lowered `TensorIndex` object to `getGlobalProducerStridedIndices` for the indexing of the given ID.

Reply:
I was wondering if it would make sense to allow an initial index map to be given to the index computation. In the case of `select`, we could initialize a map for the selected ID with the second argument of the select expression. For `index_select`, it would be a 2-step indexing like you mentioned. The first step would be indexing of the index argument, and then we could create a map of the selected ID to the lowered index tensor. Then `getGlobalProducerStridedIndices` would use this optional map to index into the selected root ID. This is similar to your approach, but maybe a little more generic?

Reply:
I agree. Working on it.

Reply:
I did some relatively big refactor to implement it in this commit: https://github.com/csarofeen/pytorch/pull/2179/commits/87fee1daf1ae848a0e475998921bb9a259393966
@naoyam Could you please take another look to see if it makes sense to you?
That commit also resolves some of your review comments, but not all of them. I will work on some other review comments in later commits in this PR.